netket/optimizer/sr/base.py

```diff
@@ -32,3 +34,96 @@ class SR:
+    """
+    S matrix base class.
+    This can either be a jnp matrix, a lazy wrapper, or anything, as long as
+    it satisfies this basic API.
+
+    An AbstractSMatrix must support the following API:
+
+    - :code:`__matmul__(y)`, meaning you must be able to do S@vec, where vec is
+        either a PyTree of parameters or it's dense ravelling.
+    - :code:`solve(y, **kwargs)`, which must solve the linear system Sx=y with
+        any method available, usually but not necessarily defined inside the
+        sr object stored inside the AbstractSMatrix.
+        This function must accept arbitrary additional arguments.
+
+
+    Additionally, you get for free:
+    - :code:`__call__(y)`, meaning you must be able to do S(vec) = S@vec.
+        This is defined by the base class so you don't need to define it.
+        This guarantees that you can pass this matrix to sparse solvers.
+    - :code:`__rtruediv__(y)`, that allows you to solve Sx=y by doing y\\S.
+    - :code:`to_dense(self)`, that will concretize the dense representation.
+        You get a slow default fallback by default, you might specify a faster
+        override.
+
+    """
+
+    sr: SR
+    """Parameters for the solution of the system."""
+
+    # PUBLIC API: METHOD TO EXTEND IF YOU WANT TO DEFINE A NEW S object
+    def __matmul__(self, vec):
+        raise NotImplementedError()
+
+    # PUBLIC API: METHOD TO EXTEND IF YOU WANT TO DEFINE A NEW S object
+    def solve(self, y: PyTree, **kwargs) -> PyTree:
+        """
+        Solve the linear system x=⟨S⟩⁻¹⟨y⟩ with the chosen iterataive solver.
+
+        Args:
+            y: the vector y in the system above.
+            kwargs: Any additional kwargs, which might or might not be used depending
+                on the specific implementation.
+
+        Returns:
+            x: the PyTree solving the system.
+            info: optional additional informations provided by the solver. Might be
+                None if there are no additional informations provided.
+        """
+        raise NotImplementedError()
+
+    # PUBLIC API: METHOD TO EXTEND IF YOU WANT TO DEFINE A NEW S object
+    def to_dense(self) -> jnp.ndarray:
+        """
+        Convert the lazy matrix representation to a dense matrix representation.s
+
+        Returns:
+            A dense matrix representation of this S matrix.
+        """
+        raise NotImplementedError()
+
+    # PUBLIC API: Only override if you want to, but there should be no need.
+    def __call__(self, vec):
+        return self @ vec
+
+    # PUBLIC API: Only override if you want to, but there should be no need.
+    def __rtruediv__(self, y):
+        x, _ = self.solve(y)
+        return x
+
+    # PUBLIC API: Only override if you want to, but there should be no need.
+    def __array__(self) -> jnp.ndarray:
+        return self.to_dense()
+
+
+@dispatch
+def SMatrix(sr: SR, vstate: object) -> AbstractSMatrix:
+    """
+    Construct the S matrix given a
```

Comment:
Given a ...?

Reply:
I should have deleted this.
It's part of other changes that I was experimenting with.